src/panoptes/pocs/focuser/focuser.py

```diff
@@ -401,51 +393,57 @@ def _autofocus(self,
             focus_range = focus_range[0]
             focus_step = focus_step[0]
 
+        # Get focus steps.
         focus_positions = np.arange(max(initial_focus - focus_range / 2, self.min_position),
                                     min(initial_focus + focus_range / 2, self.max_position) + 1,
                                     focus_step, dtype=np.int)
         n_positions = len(focus_positions)
 
-        thumbnails = np.zeros((n_positions, thumbnail_size, thumbnail_size),
-                              dtype=initial_thumbnail.dtype)
+        # Set up empty array holders
+        thumbnails = np.zeros((n_positions, thumbnail_size, thumbnail_size), dtype=initial_thumbnail.dtype)
         masks = np.empty((n_positions, thumbnail_size, thumbnail_size), dtype=np.bool)
-        metric = np.empty(n_positions)
+        metrics = np.empty(n_positions)
 
         # Take and store an exposure for each focus position.
         for i, position in enumerate(focus_positions):
             # Move focus, updating focus_positions with actual encoder position after move.
-            focus_positions[i] = self.move_to(position)
+            encoder_position = self.move_to(position)
+            focus_positions[i] = encoder_position
 
-            # Take exposure
-            focus_fn = "{}_{:02d}.{}".format(focus_positions[i], i, self._camera.file_extension)
+            focus_fn = f"{encoder_position}-{i:02d}.{self._camera.file_extension}"
             file_path = os.path.join(file_path_root, focus_fn)
 
+            # Take exposure.
             try:
-                thumbnail = self._camera.get_thumbnail(
-                    seconds, file_path, thumbnail_size, keep_file=keep_files)
+                thumbnail = self._camera.get_thumbnail(seconds, file_path, thumbnail_size, keep_file=keep_files)
             except Exception as err:
                 self.logger.error(f"Error taking image {i + 1}: {err!r}")
                 self._autofocus_error = repr(err)
                 focus_event.set()
                 raise err
 
             masks[i] = mask_saturated(thumbnail, bit_depth=self.camera.bit_depth).mask
-            if dark_thumb is not None:
-                thumbnail = thumbnail - dark_thumb
+            if dark_thumbnail is not None:
+                thumbnail = thumbnail - dark_thumbnail
             thumbnails[i] = thumbnail
 
+        self.logger.debug(f'Making master mask with binary dilation for {self._camera}')
         master_mask = masks.any(axis=0)
         master_mask = binary_dilation(master_mask, iterations=mask_dilations)
 
         # Apply the master mask and then get metrics for each frame.
         for i, thumbnail in enumerate(thumbnails):
+            self.logger.debug(f'Applying focus metric to thumbnail {i:02d}')
             thumbnail = np.ma.array(thumbnail, mask=master_mask)
```

Comment:
This is a bug. `thumbnail` will in general already have a mask, the bad pixel mask inherited from `dark_thumbnail` during dark subtraction. Instead of just replacing that with `master_mask` (which removes the bad pixel masking) we should be combining `master_mask` with the existing mask.

```suggestion
            thumbnail = np.ma.array(thumbnail, mask=np.ma.mask_or(master_mask, np.ma.getmask(thumbnail)))
```

Reply:
Wow, how did I now know about `mask_or`..Thanks!